core/embed/rust/src/micropython/buffer.rs

```diff
@@ -27,22 +32,48 @@ impl StrBuffer {
     }
 
     pub fn alloc(val: &str) -> Result<Self, Error> {
+        unsafe {
+            // SAFETY: Copying existing UTF-8 buffer of the right length.
+            Self::alloc_with(val.len(), |buffer| {
+                // SAFETY: Memory should be freshly allocated and as such cannot overlap.
+                ptr::copy_nonoverlapping(val.as_ptr(), buffer.as_mut_ptr(), buffer.len());
+                // SAFETY: Copy of valid UTF-8 string.
+                str::from_utf8_unchecked(slice::from_raw_parts(buffer.as_ptr(), buffer.len()))
+            })
+        }
+    }
+
+    /// SAFETY: The caller must ensure that `func` returns a slice that:
+    /// - spans the entire length of the given buffer, and
+    /// - contains valid UTF-8.
+    /// Note that every byte in the buffer has to be overwritten for this to
+    /// hold.
+    pub unsafe fn alloc_with(
+        len: usize,
+        func: impl FnOnce(&mut [u8]) -> &str,
+    ) -> Result<Self, Error> {
         // SAFETY:
         // We assume that if `gc_alloc` returns successfully, the result is a valid
         // pointer to GC-controlled memory of at least `val.len() + 1` bytes.
         unsafe {
-            let raw = ffi::gc_alloc(val.len() + 1, 0) as *mut u8;
+            let raw = ffi::gc_alloc(len + 1, 0) as *mut u8;
             if raw.is_null() {
                 return Err(Error::AllocationFailed);
             }
-            // SAFETY: Memory should be freshly allocated and as such cannot overlap.
-            ptr::copy_nonoverlapping(val.as_ptr(), raw, val.len());
+
+            // SAFETY: GC returns valid pointers, slice is discarded after `func`.
+            let bytes = slice::from_raw_parts_mut(raw, len);
+            let initialized = func(bytes);
+            assert_eq!(initialized.as_ptr(), raw as *const u8);
+            assert_eq!(initialized.len(), len);
```

Comment:
Maybe instead fill the whatever's after initialized with zeroes? Need to make sure uninitialized memory doesn't escape.

Reply:
is an "uninitialized" `&[u8]` slice a problem?
my gut says no because whatever content is there, it's still makes for a valid `&[u8]` data.

Reply:
It may prevent other dead objects from being garbage collected: https://github.com/micropython/micropython/blob/master/py/gc.c#L689-L694